### PR TITLE
[Android] ToolbarHandler and fixes for various page nesting scenarios

### LINKED ISSUE #1658: [Bug] Updating Page title anywhere after constructor doesn't reflect, observed while using NavigationPage
### Description
Updating Page title anywhere after constructor doesn't reflect, observed while using NavigationPage

### Steps to Reproduce

1. Create a new dotnet Maui project, use NavigationPage with a ContentPage as a startup in CreateWindow() method
2. Try to modify the Page title in any event apart from the constructor
3. Title is not reflecting in the navigation bar

### Expected Behavior
Updated title to get displayed in the navigation bar

### Actual Behavior
The updated title doesn't reflect, whatever that is set in the constructor stays as it is.

### Basic Information

- Version with issue: **.NET 6 Preview 5**
- Last known good version: NA
- IDE: **VS2019 16.11 Preview 2.0**
- Platform Target Frameworks: 
  - iOS:  
  - Android:  **Target API level 30 (Android 11)**
  - UWP:   
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

### LINKED ISSUE #2434: [Bug] [Android] OnAppearing not called when using Page navigation.
### Description
Using Maui `6.0.100-rc.2.1642`

When using a Navigation Page on Android, `OnAppearing` is not fired. It seems to only be called when the page is first created on startup.

https://user-images.githubusercontent.com/898335/132733107-d2da340b-6333-4222-a36c-74e6363c7682.mp4

### Steps to Reproduce

Set up OnAppearing on a ContentPage and put a breakpoint there. It will be hit when the page is created, but not when navigated to or back from.

This does work on on WinUI.

### LINKED ISSUE #2521: [Bug] [Android] InvalidOperationException "NavigationStack cannot be empty" when attempting back navigation on empty NavigationPage
Using Maui `6.0.100-rc.2.7147`

### Description

https://user-images.githubusercontent.com/898335/133633830-74f8637b-4993-431e-b86e-8b87f08db08d.mp4

### Steps to Reproduce

1. Create a navigation page.
2. On Android, use the back button (if available) and try to go back on this initial page.

### Expected Behavior

The app does nothing, since it's the root page.

### Actual Behavior

The app crashes here: https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Android/Navigation/NavigationManager.cs#L90

This could be handled by the user, checking if the navigation stack is empty before allowing events from the Android back button to be fired. However, IMO, that's not obvious and could lead to very easy bugs in real-world applications. I would think that if the navigation stack is empty, and the back button is called, nothing would happen.

### LINKED ISSUE #3217: [Bug] ToolbarItem handler is not getting called
### Description

ToolbarItem handler is not getting called

### Steps to Reproduce

1. Create a new content page derived class.
2. Create a new ToolbarItem, supply click handler, and add to the buttons collection

### Expected Behavior

The toolbar item handler is getting called on item click

### Actual Behavior

The toolbar item handler is not getting called

### Basic Information

- Version with issue: 6.0.100-rc.2.21505.57
- Last known good version: NA
- IDE: VS2022 - Preview 5
- Platform Target Frameworks: iOS, Android, Windows

### Screenshots

image

### Reproduction Link

```
// contstructor
ToolbarItem mauiButton = new ToolbarItem
{
    Text = "Add",
    Order = ToolbarItemOrder.Primary,
    Priority = 0
};
mauiButton.Clicked += MauiButtonClicked;
this.ToolbarItems.Add(mauiButton);
```

### Workaround

None is found

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

- Sets the root container of WindowHandler.Android.cs to be a CoordinatorLayout. 
- Moves all the code for handling the toolbar into a Toolbar class and ToolbarHandler.
  - This allows for us to put all the logic for the rules of the toolbar in the xplat code and then the native code is just in charge of applying the visual changes. 
- Fixes the scenario if you have Navigation Page => TabbedPage
- Fixes the scenario if you have TabbedPage => with Navigation Pages
- Adjusted appearing on page to only fire after the parent window has been set. This starts the process of streamlining the ordering of events. Right now in XF parentset/appearing fire in different orders in no consistent way. Appearing should fire after ParentSet so that users can utilize the visual tree hierarchy. 
  - I needed an accurate appearing event in order to fire toolbar changed events. For cases when switching NavigationPages inside tabs or pushing new pages onto the stack 

### Additions made ###

```C#
public class Toolbar : Microsoft.Maui.IElement
{
	public IEnumerable<ToolbarItem> ToolbarItems
	public bool IsVisible
	public bool HasBackStack
	public bool BackButtonVisible 
	public double? BarHeight 
	public string BackButtonTitle
	public ImageSource TitleIcon 
	public Color BarBackgroundColor 
	public Brush BarBackground 
	public Color BarTextColor
	public Color IconColor 
	public string Title 
	public VisualElement TitleView 
}
```

```C#
public class ToolbarHandler : ElementHandler<Toolbar, MaterialToolbar>
```

```C#
// Renamed from NavigationManager
public class StackNavigationManager
```

### CODE REVIEW

File: src/Controls/src/Core/Handlers/Toolbar/ToolbarHandler.Android.cs [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+using AndroidX.AppCompat.Graphics.Drawable;
+using AndroidX.DrawerLayout.Widget;
+using Google.Android.Material.AppBar;
+using Microsoft.Maui.Controls.Platform;
+using Microsoft.Maui.Graphics;
+using Microsoft.Maui.Handlers;
+
+namespace Microsoft.Maui.Controls.Handlers
+{
+	internal class ToolbarHandler : ElementHandler<Toolbar, MaterialToolbar>
```
Comment (PureWeen) [inline src/Controls/src/Core/Handlers/Toolbar/ToolbarHandler.Android.cs:21]: All the commented out code in here is related to FlyoutPage and once we wire that all up

File: src/Controls/src/Core/Platform/Android/Shell/ShellPageContainer.cs [hunk 2]
```diff
@@ -37,5 +42,14 @@ protected override void OnLayout(bool changed, int l, int t, int r, int b)
 			if (changed && Child.NativeView is AView aView)
 				aView.Layout(0, 0, width, height);
 		}
+
+		protected override void OnMeasure(int widthMeasureSpec, int heightMeasureSpec)
```
Comment (PureWeen) [inline src/Controls/src/Core/Platform/Android/Shell/ShellPageContainer.cs:46]: This is all because I deleted PageContainer

**Comment by PureWeen** [comment]
Related Issues
- https://github.com/dotnet/maui/issues/2804

File: src/Core/src/Platform/Android/HandlerExtensions.cs [hunk 3]
```diff
@@ -36,10 +36,13 @@ public static AView ToNative(this IElement view, IMauiContext context)
...
 				handler.SetVirtualView(view);
 
-			if (((INativeViewHandler)handler).NativeView is not AView result)
-				throw new InvalidOperationException($"Unable to convert {view} to {typeof(AView)}");
+			if (handler is INativeViewHandler nvh && nvh.NativeView is AView)
+				return nvh.NativeView;
 
-			return result;
+			if (handler.NativeView is AView result)
+				return result;
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Core/src/Platform/Android/HandlerExtensions.cs:43]: No, The toolbarhandler is just an ElementHandler and that's it

I tried implementing from `INativeViewHandler` but `INativeViewHandler` implements `IViewHandler`

If I implement `IViewHandler` then my virtual view needs to be an `IView` but I don't want my Toolbar.cs to be an `IView` it's just an `IElement`  😢 

So we might need to convert `INativeViewHandler` to `INativeElementHandler` :-)

**Review by hartez — APPROVED** [review]

**Pull request merged**